Add Contains to parser tests.

srh opened the pull request:
This adds Contains to a bunch of parser tests, to make sure they're failing where they're supposed to be failing.

Also fixes some tests which had spurious parse errors.

In parser_test.go I added some TODO comments for test cases I didn't understand.

My advice for reviewing this is to focus on the TODO's in the diff.  The rest of the tests were straightforward enough.

Connected to #1229.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/1307)
<!-- Reviewable:end -->

pawanrawal commented:
Thanks for taking the pain to do this. It's really useful and something that we should do from now on in other tests as well. 

---

Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions, some commit checks broke.

---

*[gql/parser_test.go, line 177 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqee7e-2Ht49-DrDzk7:-Kqee7e-2Ht49-DrDzk8:bod30or) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L177)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.NoError(t, err)
> 	// TODO: This used to error.  What was this supposed to test?
> ```

It should error because of `uid()`. I think it throws an error while processing in `query.go` but should be handled here.

---

*[gql/parser_test.go, line 214 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqeeT2LMZiUGlsTyA9V:-KqeeT2LMZiUGlsTyA9W:b-b3ga01) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L214)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err)
> 	// TODO: What is this testing?  "alice-in-wonderland" having "-in" is the source of error here
> ```

`uid()` function should only accept arguments which can be parsed as a uint64 or a variable defined in another block hence this should error. 

---

*[gql/parser_test.go, line 254 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqeepmrNygXIhnvMXPS:-KqeepmrNygXIhnvMXPT:b3gzjxj) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L254)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err)
> 	// TODO: This now says, "Unexpected comma before )."
> ```

Probably a bad error message but I think error is in `sumvar()`

---

*[gql/parser_test.go, line 639 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefEW_ym7smc7UzWDB:-KqefEW_ym7smc7UzWDC:bmbnyrd) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L639)):*
> ```Go
> 	query := `
> 	{
> 		me(func: uid( val(L), orderasc: val(n)) {
> ```

Should error out because the syntax should be
```
me(func: uid(L), orderdesc: val(n)) {
  ....
}
```

---

*[gql/parser_test.go, line 1149 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefoCHIh6yKPsTQtV-:-KqefoCHIh6yKPsTQtV0:bwyodxr) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L1149)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err)
> 	// TODO: This one complains about "Unrecognized character in lexText: U+005D ']'" which is a
> ```

This should actually fail at `[` because we would try to parse as a uint64 but fail. 

---

*[gql/parser_test.go, line 1982 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqegetmZ0z-QqGW438V:-KqegetmZ0z-QqGW438W:b1wdyxq) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L1982)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err, "Expected value for variable $c")
> 	// TODO: Complains about '\"'.
> ```

I think the problem is having a newline. We should remove newlines in the query. If you observe the tests that pass, they don't have a newline. This is because we do a `JSON.Unmarshal` and values in JSON can't have newlines. 
```
"query": "query testQuery($a: bool , $b: float! = 3){root(func: uid( 0x0a) {name(first: $b)){english}}}",
```

---

*[gql/parser_test.go, line 2128 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqehbvDnqvfACQjxAj0:-KqehbvDnqvfACQjxAj1:b8brlt4) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2128)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.NoError(t, err)
> 	// TODO: This was an error, but I don't think it was supposed to be a lexing-on-@ error.  Now
> ```

Not sure what it was supposed to test. I think there is already a test which fails on lexing `@` error so this looks repetitive. 

---

*[gql/parser_test.go, line 2449 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqei-DjIMKAxc1RadYt:-Kqei-DjIMKAxc1RadYu:bsqygmd) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2449)):*
> ```Go
> 	require.Error(t, err)
> 	require.Contains(t, err.Error(), "Unclosed action")
> 	// TODO: What's an unclosed action?
> ```

It's complaining about a missing `}` and the action here refers to `query` I think.

---

*[gql/parser_test.go, line 2466 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqeiEf4NjZAcTO3R10F:-KqeiEf4NjZAcTO3R10G:b-lbtwse) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2466)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err)
> 	// TODO: Says "expecting a colon".  Should be me(func: ...).  Then it complains about "name"
> ```

From the name I think this is supposed to test the missing `func` keyword error at root. There should be another test which tests that `name` shouldn't be quoted (this restriction is a recent addition). 

---

*[gql/parser_test.go, line 2877 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqeje-MqMONVF_hiz1z:-Kqeje-MqMONVF_hiz2-:bgcbmsk) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2877)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err)
> 	// TODO: We get 'Expected 3 periods ("..."), got 2.'.
> ```

I know whats happening here.
Syntax for language lists is `name@hi:cn:.`(one dot at the end) and syntax for fragments is 
`...frag`(three dots at beginning). So after language list has consumed one dot, we assume its a fragment and when we can't parse three dots we throw an error.

I think we should peek while parsing language lists and ensure that we don't get more than one dot and throw an error there like `Expected 1 period, got more`. 

---

*[gql/parser_test.go, line 2975 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqelT9eqR1fnICqo7TS:-KqelT9eqR1fnICqo7TT:bd5arwr) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2975)):*
> ```Go
> func TestLangsFunctionMultipleLangs(t *testing.T) {
> 	// TODO: vvv What?  Touches global variables??
> 	schema.ParseBytes([]byte("scalar descr: string @index(fulltext) ."), 0)
> ```

I don't think this is needed here because the schema is not checked while parsing. Not sure why this was added.

---

*[gql/parser_test.go, line 3567 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqelBguXd1c0T22FixT:-KqelBgvKtAzyN75gJVB:b-ri69gg) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L3567)):*
> ```Go
> 	require.Error(t, err)
> 	require.Contains(t, err.Error(), "variables are not allowed in facets filter")
> 	// TODO: This used "id(L)" before.  What is that?
> ```

That's incorrect syntax. Thanks for correcting. 

---

*[gql/parser_test.go, line 3742 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqelhqdStpjiUDKlDoc:-Kqelhqeoj8fgR704npG:bibvjou) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L3742)):*
> ```Go
> 	_, err := Parse(Request{Str: query, Http: true})
> 	require.Error(t, err)
> 	// TODO: I'm not sure what this test is supposed to be testing.
> ```

We pass `a`, `b` and `c` as variables to uid function but `b` and `c` are not defined but the syntax for specifying those also looks incorrect. It should be
`func: uid($a,$b,$c)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-:-KqemuhzgcQIYlgPHzmS:burcccr)*
<!-- Sent from Reviewable.io -->

srh commented:
Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions, some commit checks broke.

---

*[gql/parser_test.go, line 177 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqee7e-2Ht49-DrDzk7:-Kqf3NCijIs4pFsbs0bs:b8mgkgz) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L177)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

It should error because of `uid()`. I think it throws an error while processing in `query.go` but should be handled here.
</blockquote></details>

Okay.  I don't see why `uid` couldn't take zero parameters if it takes a variable number of them.

---

*[gql/parser_test.go, line 214 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqeeT2LMZiUGlsTyA9V:-Kqew62-slARddIxpULl:b-o18ka9) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L214)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

`uid()` function should only accept arguments which can be parsed as a uint64 or a variable defined in another block hence this should error. 
</blockquote></details>

OK, fixed.

---

*[gql/parser_test.go, line 254 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqeepmrNygXIhnvMXPS:-KqewYtIMuV-3gDr66L4:bv2qeg1) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L254)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Probably a bad error message but I think error is in `sumvar()`
</blockquote></details>

That's right.  Changed the error message to be a bit more general.

---

*[gql/parser_test.go, line 639 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefEW_ym7smc7UzWDB:-KqewsTOCYh0c3ts51KB:b-kc8vj4) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L639)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Should error out because the syntax should be
```
me(func: uid(L), orderdesc: val(n)) {
  ....
}
```
</blockquote></details>

If I do `val(L)` instead of `uid(L)`, it passes.

---

*[gql/parser_test.go, line 1149 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefoCHIh6yKPsTQtV-:-KqexJ8s-5VRsNDi9EF8:b-58toyf) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L1149)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

This should actually fail at `[` because we would try to parse as a uint64 but fail. 
</blockquote></details>

If I try `uid( [0x1, 0x1, 2, 3, 0x34])` it passes.

---

*[gql/parser_test.go, line 1982 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqegetmZ0z-QqGW438V:-Kqf-3RZABMRUGjagkK1:bjb6oxv) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L1982)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I think the problem is having a newline. We should remove newlines in the query. If you observe the tests that pass, they don't have a newline. This is because we do a `JSON.Unmarshal` and values in JSON can't have newlines. 
```
"query": "query testQuery($a: bool , $b: float! = 3){root(func: uid( 0x0a) {name(first: $b)){english}}}",
```
</blockquote></details>

Yeah, that's right.  Tab characters are also a problem.

---

*[gql/parser_test.go, line 2128 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqehbvDnqvfACQjxAj0:-Kqf-ut8l7clUVjXShMf:b-9mt3tr) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2128)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Not sure what it was supposed to test. I think there is already a test which fails on lexing `@` error so this looks repetitive. 
</blockquote></details>

Removed it then.

---

*[gql/parser_test.go, line 2449 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqei-DjIMKAxc1RadYt:-Kqf0Gw_8TGvEjwkj6jP:b-qj3zzo) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2449)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

It's complaining about a missing `}` and the action here refers to `query` I think.
</blockquote></details>

I didn't notice that.  If I add '}' it successfully parses -- but shouldn't it fail because of @filtererr?

---

*[gql/parser_test.go, line 2466 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqeiEf4NjZAcTO3R10F:-Kqf0h1Jrbg-U5hNWudU:b-wl0gcl) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2466)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

From the name I think this is supposed to test the missing `func` keyword error at root. There should be another test which tests that `name` shouldn't be quoted (this restriction is a recent addition). 
</blockquote></details>

OK, I added a second test.

---

*[gql/parser_test.go, line 2877 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqeje-MqMONVF_hiz1z:-Kqf0rrwTpcyRbDbXKf4:blfcm0p) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2877)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I know whats happening here.
Syntax for language lists is `name@hi:cn:.`(one dot at the end) and syntax for fragments is 
`...frag`(three dots at beginning). So after language list has consumed one dot, we assume its a fragment and when we can't parse three dots we throw an error.

I think we should peek while parsing language lists and ensure that we don't get more than one dot and throw an error there like `Expected 1 period, got more`. 
</blockquote></details>

Could lexing handle groups of dots?

---

*[gql/parser_test.go, line 2975 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqelT9eqR1fnICqo7TS:-Kqf1-XTk5Sgl2gFR-RI:b-76aero) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2975)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

I don't think this is needed here because the schema is not checked while parsing. Not sure why this was added.
</blockquote></details>

That's right, it's not.

---

*[gql/parser_test.go, line 3742 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqelhqdStpjiUDKlDoc:-Kqf2SgggDFLFMz9Mdob:bqmntjk) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L3742)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

We pass `a`, `b` and `c` as variables to uid function but `b` and `c` are not defined but the syntax for specifying those also looks incorrect. It should be
`func: uid($a,$b,$c)`
</blockquote></details>

Then we get "Invalid use of comma"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[gql/parser_test.go, line 177 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqee7e-2Ht49-DrDzk7:-KqfPwCALS2QtkCUf7zI:bypqjvf) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L177)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

Okay.  I don't see why `uid` couldn't take zero parameters if it takes a variable number of them.
</blockquote></details>

It can take zero parameters though it won't make much sense because the result would always be empty.

---

*[gql/parser_test.go, line 639 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefEW_ym7smc7UzWDB:-KqfQKGunI9QJMqY_8kn:b-k81294) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L639)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

If I do `val(L)` instead of `uid(L)`, it passes.
</blockquote></details>

Do you mean this passes? If it does then that's a bug.
```
me(func: val(L), orderdesc: val(n)) {
  ....
}
```

---

*[gql/parser_test.go, line 1149 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefoCHIh6yKPsTQtV-:-KqfQ_qzDaiy373CYQjy:b-opnxvn) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L1149)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

If I try `uid( [0x1, 0x1, 2, 3, 0x34])` it passes.
</blockquote></details>

Hmm, I think this is unexpected behavior. Earlier syntax for specifying ids was
`me(id: [1, 2, 3])` and I suppose we didn't get rid of the handling for square brackets in the parser when we moved to the new syntax with a uid function which takes multiple arguments hence no need for `[]`.

---

*[gql/parser_test.go, line 2449 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqei-DjIMKAxc1RadYt:-KqfRFL0fPsbkAXssK6s:bqlmu79) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2449)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

I didn't notice that.  If I add '}' it successfully parses -- but shouldn't it fail because of @filtererr?
</blockquote></details>

Probably you were referring to `filtererr` in the test above. 

I think in `getRoot` it would be good to check that the key should be one of `func`, `orderasc`, `orderdesc` , `first` , `after`, `offset`. You can add a TODO for this and I can do it later.

---

*[gql/parser_test.go, line 2877 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqeje-MqMONVF_hiz1z:-KqfTEZtRHOZGn1VegNT:buf03qs) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2877)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

Could lexing handle groups of dots?
</blockquote></details>

Possibly, will have to check code. 

---

*[gql/parser_test.go, line 3742 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqelhqdStpjiUDKlDoc:-KqfT2rWktnVnZD4KC6I:b-602o2g) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L3742)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

Then we get "Invalid use of comma"
</blockquote></details>

Let the TODO remain. I can check this later, need to see whats going on here. 

---

*[gql/parser_test.go, line 2431 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqfRVAz6QJg4T4BtEwR:-KqfRVB--jvTgsYZgFr2:bmt4vk1) ([raw file](https://github.com/dgraph-io/dgraph/blob/b82f9bd2464156e4aab3be9d371002c5a34adb9f/gql/parser_test.go#L2431)):*
> ```Go
> 	require.NoError(t, err)
> 	// require.Contains(t, err.Error(), "Unclosed action")
> 	// TODO: Should this fail to parse?
> ```

Yeah, this should fail. Probably another bug if it doesn't fail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307)*
<!-- Sent from Reviewable.io -->

srh commented:
I made issue https://github.com/dgraph-io/dgraph/issues/1311 for you.

---

Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[gql/parser_test.go, line 639 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefEW_ym7smc7UzWDB:-KqfXZwYl29HpE19WGLO:b-exapc6) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L639)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Do you mean this passes? If it does then that's a bug.
```
me(func: val(L), orderdesc: val(n)) {
  ....
}
```
</blockquote></details>

Yes.  I'll file an issue.

---

*[gql/parser_test.go, line 1149 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqefoCHIh6yKPsTQtV-:-KqfZnll1oCLOGUtzhyt:b-6jbnry) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L1149)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Hmm, I think this is unexpected behavior. Earlier syntax for specifying ids was
`me(id: [1, 2, 3])` and I suppose we didn't get rid of the handling for square brackets in the parser when we moved to the new syntax with a uid function which takes multiple arguments hence no need for `[]`.
</blockquote></details>

OK.  I'll file an issue.  Maybe the same one.

---

*[gql/parser_test.go, line 2449 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-Kqei-DjIMKAxc1RadYt:-Kqf_OxewdKRssdIHVbv:b-hj3ub9) ([raw file](https://github.com/dgraph-io/dgraph/blob/47b85270b266026e38b8a55ff27bbba501368240/gql/parser_test.go#L2449)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Probably you were referring to `filtererr` in the test above. 

I think in `getRoot` it would be good to check that the key should be one of `func`, `orderasc`, `orderdesc` , `first` , `after`, `offset`. You can add a TODO for this and I can do it later.
</blockquote></details>

I'll add that.

---

*[gql/parser_test.go, line 2431 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-KqfRVAz6QJg4T4BtEwR:-KqfZzQDSt3xQN6zIm3W:bfhrmn) ([raw file](https://github.com/dgraph-io/dgraph/blob/b82f9bd2464156e4aab3be9d371002c5a34adb9f/gql/parser_test.go#L2431)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Yeah, this should fail. Probably another bug if it doesn't fail.
</blockquote></details>

I'll include it in the issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-:-KqfbrA97UQsIo-2WL7u:b-7hvi1y)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1307#-:-Kqfhs5pAoawCTxp4MuE:boophy0)*
<!-- Sent from Reviewable.io -->

srh commented:
In master as of 89dc3dd197cc13b4434ccc49f1a7744ebd8660ef.
